### PR TITLE
update the last_used_utc on the webauthn record IDP-690

### PR DESCRIPTION
**Pull request by briskt**
### Fixed
- Update the `last_used_utc` column on a webauthn record when it is used. [IDP-690](https://itse.youtrack.cloud/issue/IDP-690)
- Removed unused test code.

---

### Feature PR Checklist
- [x] Documentation (README, local.env.dist, etc.)
- [ ] Unit tests created or updated
- [x] Run `make composershow`

_Code and test structure does not lend itself to testing the verification of a webauthn. It may be possible using the U2F simulator. I am not familiar with how to use that, though._

### CODE REVIEW

File: application/features/bootstrap/MfaContext.php [hunk 1]
```diff
@@ -262,14 +262,6 @@ public function iRequestToVerifyTheWebauthnMfaRegistrationWithALabelOf($label)
         $this->iRequestTheResourceBe('/mfa/' . $this->mfa->id . '/verify/registration', 'created');
     }
 
-    /**
```
Comment (mtompset) [inline application/features/bootstrap/MfaContext.php:265]: Why was this removed?

Comment (briskt) [reply application/features/bootstrap/MfaContext.php:265]: I was trying to write a test that would use this, but failed. In the process, I noticed that it was not being used.

**Comment by jason-jackson** [comment]
Looks like you should be able to use the u2fsimulator, it's using the `silintl/serverless-mfa-api-go:develop` 

Example usages:
```
$this->callU2fSimulator('/webauthn/login', 'created', $user, $mfa->external_uuid);
$this->callU2fSimulator('/webauthn/login', 'updated', $user, $mfa->external_uuid);

$this->callU2fSimulator('/webauthn/register', 'created', $user, $mfa->external_uuid);
$this->callU2fSimulator('/webauthn/register', 'updated', $user, $mfa->external_uuid);

$this->callU2fSimulator('/webauthn/user', 'deleted', $user, $mfa->external_uuid);
$this->callU2fSimulator('/webauthn/credential/'.$mfa->id, 'deleted', $user, $mfa->external_uuid);  // I could be wrong about this one
```

I'd have to look at proper order for calling as it's been a while since I worked with webauthn.

**Comment by briskt** [comment]
Thanks, @jason-jackson. That's a step in the right direction. What I needed was the valid data that needs to be passed to the id-broker in an /mfa/verify call.

File: application/common/models/MfaWebauthn.php [hunk 2]
```diff
@@ -92,4 +92,17 @@ public static function createWebauthn(Mfa $mfa, string $keyHandleHash, string $l
 
         return $webauthn;
     }
+
+    public function setLastUsed()
+    {
+        $this->last_used_utc = MySqlDateTime::now();
+        if (! $this->save()) {
```
Comment (mtompset) [inline application/common/models/MfaWebauthn.php:99]: This saves more than just last_used_utc. You could limit it to just last_used_utc. But this works.

**Review by mtompset — APPROVED** [review]
I'm assuming someone has tested that it runs, because it eyeballs fine.

**Comment by briskt** [comment]
> I'm assuming someone has tested that it runs, because it eyeballs fine.

I was planning on testing it on the staging environment.

File: application/common/models/MfaWebauthn.php (hunk 2 above)
Comment (briskt) [reply application/common/models/MfaWebauthn.php:99]: What's the syntax for that? I could look it up, but I'm sure you can whip it right out.

Comment (mtompset) [reply application/common/models/MfaWebauthn.php:99]: $this->save(true, ['last_used_utc']) -- still runs the validation, but only validate and save last_used_utc.

Comment (briskt) [reply application/common/models/MfaWebauthn.php]: Thank you very much.

File: application/common/components/MfaBackendWebAuthn.php [hunk 3]
```diff
@@ -158,6 +158,9 @@ public function verify(int $mfaId, $value, string $rpOrigin = '', string $verify
             if ($webauthnCount < 1) {
                 throw new NotFoundHttpException("No MFA Webauthn record found for MFA ID: " . $mfa->id, 1659637860);
             }
+
+            $mfa->setLastUsed();
```
Comment (forevermatt) [inline application/common/components/MfaBackendWebAuthn.php:162]: It looks like this is recording a last-used date before verifying that the response was valid. Is that correct?

Comment (briskt) [reply application/common/components/MfaBackendWebAuthn.php:162]: Yes, it does look like I got that wrong. Thanks for spotting it.

Comment (briskt) [reply application/common/components/MfaBackendWebAuthn.php:162]: What's worse, as I got to looking at it further, I'm not sure it's going to work at all. I tried to call `setLastUsed` on an `Mfa` object, but I defined the function on `MfaWebauthn`. It appears that the validation is delegated to the MFA API, which doesn't give any indication as to which webauthn instance passes validation.
